File: src/org/bds/executioner/TaskLogger.java
package org.bds.executioner;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.io.Serializable;
import java.util.HashSet;

import org.bds.BdsLog;
import org.bds.data.Data;
import org.bds.data.DataS3;
import org.bds.task.Task;
import org.bds.util.Gpr;

/**
 * TaskLogger: This class adds entries to the `bds.pid.$PID` log file.
 *
 * Background: bds commmnad consists of two parts:
 * 	- `bds-exec`: A binary executable programmed in Go
 * 	- `bds-java`: A java program (this code)
 *
 * The `bds-exec` controls execution of `bds-java` and catches any errors
 * or exceptions.
 *
 * When Java process dies (e.g. die to `Ctrl-C`, fatal error, uncaught
 * exception, etc.), 'bds go' takes control and cleans up. This 'clean up'
 * includes managing several different types of entities, for example:
 * 	- deleting stale files: For example if a task fails, the tasks' output file are in undefined state, so we must clean them up
 * 	- kill processes: If the user performs a `Ctrl-C`, bds tries to clean up all running process, e.g. `task` or `sys` commands that are currently running locally
 *  - kill cluster jobs: Same as before, but for clusters (i.e. kill all `task` running in a cluster when a `Ctrl-C` is received)
 *  - terminate cloud instances: Same as before, but for cloud (i.e. kill all `task` running in a cloud when a `Ctrl-C` is received)
 *  - delete cloud queues: Communication queues for cloud processes must be cleaned up
 *
 * The ways this is done is simply by registering all these entities in a
 * file. This file is usually named `bds.pid.$PID` (where PID is the local
 * process ID). bds-exec created the file name and passes the name of the
 * file to bds-java using the (hidden) `-pid` command line option.
 *
 * This class takes care of registering all the entities that need to be logged
 * in that file. The format is one line per entry:
 * ```
 * entityId \t {'+','-'} \t command_to_delete
 * ```
 * where:
 * 	- entityId: Something that uniquely denotes the entity to remove (process ID, cluster job ID, file absolute path, instance ID, etc.)
 *  - `{+,-}`: This can be either
 *		- '+' indicates where we are adding the entity to be cleaned up
 *		- '-' indicates where we removing the entity (it has already been cleaned up by `bds-java`, so `bds-exec` doesn't need to take care of this entry (e.g. a finished task)
 *	- command: A command the has to be executed in order to clean up (e.g. `rm` to delete a file, `scancel` in s slurm cluster, etc.). Internal commands (e.g. deleting a local file, killing a local process) start with '@'
 *
 * When `bds-exec` takes control (i.e. when `bds-java` dies), it will add to a dictionary (key
 * is `entityId`, value is `command`) all entries having a '+' and delete from the dictionary all
 * `enityId` having a '-'. Then it will execute all `command` to delete the remaining entries.
 *
 * @author pcingola
 */
public class TaskLogger implements Serializable, BdsLog {

	private static final long serialVersionUID = -7712445468457053526L;

	public static final String CMD_REMOVE_FILE = "@rm";
	public static final String CMD_REMOVE_FILE_AWS_S3 = "@aws_s3_rm";
	public static final String CMD_KILL = "@kill";

	protected String fileName;
	protected HashSet<String> ids;

	public TaskLogger(String fileName) {
		if (fileName == null) throw new RuntimeException("Cannot initialize using a null file!");
		this.fileName = fileName;
		ids = new HashSet<>();
		debug("Creating Task logger, file: '" + fileName + "'");
	}

	public synchronized void add(String id, String command) {
		debug("Adding id: '" + id + "', command: '" + command + "'");
		append(createEntry(id, true, command));
	}

	/**
	 * Add a task and the corresponding executioner
	 */
	public synchronized void add(Task task, Executioner executioner) {
		debug("Adding task: '" + task.getId() + "'");
		StringBuilder lines = new StringBuilder();

		// Add pid
		String pid = task.getPid();
		ids.add(pid);

		// Append process PID
		// Prepare command
		StringBuilder cmdsb = new StringBuilder();
		String[] osKillCommand = executioner.osKillCommand(task);
		if (osKillCommand != null) {
			for (String c : executioner.osKillCommand(task))
				cmdsb.append(" " + c);
		}
		String cmd = cmdsb.toString().trim();

		// Append  entry
		lines.append(createEntry(task.getPid(), true, cmd));

		// Append task output files.
		// Note: If this task does not finish (e.g. Ctrl-C), we have to remove these files.
		//       If the task finished OK, we mark them not to be removed
		if (task.getOutputs() != null) {
			for (Data file : task.getOutputs())
				if (file.isRemote()) {
					// Remote file: Delete local copy (it's probably corrupted by the unfinished task
					lines.append(createEntry(file.getLocalPath(), true, CMD_REMOVE_FILE));

					// Can we delete the remote copy of the file?
					switch (file.getDataType()) {
					case S3:
						// We use a custom format "region,bucket,key" to make it easier to parso in bds-exec
						DataS3 d = (DataS3) file;
						String csv = d.getRegion() + ',' + d.getBucket() + ',' + d.getKey();
						lines.append(createEntry(csv, true, CMD_REMOVE_FILE_AWS_S3));
						break;

					case HTTP:
						// Cannot delete HTTP files, these should not be "output"
						break;

					case SFTP:
					case FTP:
						// Note: bds-exec doesn't handle FTP files
						break;

					case TASK:
						// This is not a real "file" data type
						break;

					default:
						throw new RuntimeException("Unhandled data type: '" + file.getDataType() + "', this should never happen!");
					}
				} else {
					// Local file, use remove command
					lines.append(createEntry(file.getAbsolutePath(), true, CMD_REMOVE_FILE));
				}
		}

		// Append all lines to file
		append(lines.toString());
	}

	/**
	 * Append a string to the pidFile
	 */
	protected void append(String str) {
		try {
			debug("Appending to Task logger file '" + fileName + "', lines:\n" + Gpr.prependEachLine("\t\t|", str));
			PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter(fileName, true)));
			out.print(str);
			out.close(); // We need to flush this as fast as possible to avoid missing PID values in the file
		} catch (Exception e) {
			throw new RuntimeException("Error appending information to file '" + fileName + "'\n", e);
		}
	}

	/**
	 * Create a log entry (a line in the log file)
	 * @param id : Item ID (e.g. file path, process ID, instance ID, etc)
	 * @param add : Add or remove
	 * @param command : Command line to remove item
	 */
	protected String createEntry(String id, boolean add, String command) {
		char sign = add ? '+' : '-';
		String line = id + "\t" + sign + "\t" + command;
		debug("Creating entry: '" + line + "'");
		return line + '\n';
	}

	public HashSet<String> getPids() {
		return ids;
	}

	public synchronized void remove(String id) {
		debug("Removing id: '" + id + "'");
		append(createEntry(id, false, ""));
	}

	/**
	 * Remove a task
	 */
	public synchronized void remove(Task task) {
		debug("Removing task: '" + task.getId() + "'");
		// Remove PID
		String pid = task.getPid();
		ids.remove(pid);

		StringBuilder lines = new StringBuilder();

		// Append process PID
		lines.append(createEntry(task.getPid(), false, ""));

		// Append task output files.
		if (task.getOutputs() != null) {
			for (Data file : task.getOutputs())
				lines.append(createEntry(file.url(), false, ""));
		}

		// Append all lines to file
		append(lines.toString());
	}
}

File: go/bds/src/exec/taskLogger.go
package exec

import (
	"aws"
	"bufio"
	"fileutil"
	"log"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"syscall"
	"tmpfile"
)

// Command indicating to remove file (taskLogger file)
const CMD_REMOVE_FILE = "@rm"	// Command to delete a local file
const CMD_KILL = "@kill"	// Command to kill a local process
const CMD_AWS_DELETE_QUEUE_COMMAND = "@aws_sqs_delete_queue"	// Command to delete a queue
const CMD_AWS_TERMINATE_INSTANCE = "@aws_ec2_terminate"	// Command to terminate an EC@ instance on AWS
const CMD_AWS_S3_DELETE_FILE = "@aws_s3_rm"

func (be *BdsExec) createTaskLoggerFile() {
	prefix := "bds.pid." + strconv.Itoa(syscall.Getpid())
	pidTmpFile, err := tmpfile.TempFile(prefix)
	if err != nil {
		log.Fatal(err)
	}
	be.TaskLoggerFile = pidTmpFile
	if DEBUG {
		log.Printf("Debug createTaskLoggerFile: Creating file '%s'\n", be.TaskLoggerFile)
	}
}

/*
	Perform final clean up: Parse taskLoggerFile
		i) Send kill signal to all process groups that have not been marked as 'finished'
		ii) Run commands neede to deallocate processes (e.g. cluster)
		iii) Remove stale output files form unfinshed tasks

	File format:
		"pid \t {+,-} \n"

	where '+' inidicates the process was started and '-' that
	the process finished. So all pid that do not have a '-' entry
	must be killed.
*/
func (be *BdsExec) TaskLoggerCleanUpAll() {
	if DEBUG {
		log.Printf("Debug taskLoggerCleanUpAll: Start '%s'\n", be.TaskLoggerFile)
	}

	// Parse task logger file
	defer os.Remove(be.TaskLoggerFile) // Make sure the PID file is removed
	cmds := be.taskLoggerParseFile()

	// Kill all pending processes
	be.taskLoggerProcess(cmds)
}

/*
 Parse taskLogger file
 Format is one line per entry:
 	id \t {"+", "-"} \t command

  Returns a hash of commands to run, to remove the resources
*/
func (be *BdsExec) taskLoggerParseFile() map[string]string {
	cmds := make(map[string]string)

	// Open file and parse it
	file, err := os.Open(be.TaskLoggerFile)
	if err != nil {
		log.Printf("Error taskLoggerParseFile: Cannot open TaskLogger file '%s' (PID: %d)\n", be.TaskLoggerFile, syscall.Getpid())
		return cmds
	}
	defer file.Close() // Make sure the file is deleted

	// Read line by line
	if DEBUG {
		log.Printf("Debug taskLoggerParseFile: Parsing taskLogger file '%s'\n", be.TaskLoggerFile)
	}
	reader := bufio.NewReader(file)
	for {
		line, err := fileutil.ReadLine(reader)
		if err != nil {
			break
		}
		recs := strings.Split(line, "\t")

		pid := recs[0]
		addDel := recs[1]

		// Add or remove from map
		switch addDel {
			case "-":
				delete(cmds, pid)
				if DEBUG {
					log.Printf("Debug taskLoggerParseFile: Removing id '%s', add/del: '%s'\n", pid, addDel)
				}
			case "+":
				if len(recs) > 2 && len(recs[2]) > 0 {
					cmd := recs[2]
					cmds[pid] = cmd
					if DEBUG {
						log.Printf("Debug taskLoggerParseFile: Adding id '%s', add/del '%s', cmd '%s'\n", pid, addDel, cmd)
					}
				} else {
					log.Printf("Error taskLoggerParseFile: Invalid line, cmd field not found, line: '%s'\n", cmds, line)
				}
			default:
				log.Printf("Error taskLoggerParseFile: Invalid addDel field '%s', line: '%s'\n", addDel, line)
		}
	}

	return cmds
}

/*
  Process commands in hash to remove resources
  Some "internal" commands are executed directly (e.g. remove files, kill local processes, etc.)
  All "internal" command names start with an '@' character
*/
func (be *BdsExec) taskLoggerProcess(cmds map[string]string) {
	runCmds := make(map[string]string)
	var ec2 *aws.AwsEc2
	var s3 *aws.AwsS3
	for pid, cmd := range cmds {
		switch cmd {
			case CMD_KILL:
				if VERBOSE {
					log.Printf("Info: Killing PID '%s'\n", pid)
				}
				pidInt, _ := strconv.Atoi(pid)
				be.KillProcessGroup(pidInt) // No need to run a command, just kill local porcess group
			case CMD_REMOVE_FILE:
				// This is a file to be removed, not a command
				if VERBOSE {
					log.Printf("Info: Deleting file '%s'\n", pid)
				}
				os.Remove(pid)
			case CMD_AWS_DELETE_QUEUE_COMMAND:
				if VERBOSE {
					log.Printf("Info: Deleting AWS SQS queue '%s'\n", pid)
				}
				awssqs, err := aws.NewSqs(pid, "")
				if err != nil {
					log.Printf("Error: Creating AWS SQS client for queue '%s', error: %s\n", pid, err)
				} else {
					err := awssqs.Delete() // Delete queue
					if err != nil {
						log.Printf("Error: Deleting AWS SQS queue '%s', error: %s\n", pid, err)
					}
				}
			case CMD_AWS_S3_DELETE_FILE:
				if VERBOSE {
					log.Printf("Info: Deleting AWS S3 file '%s'\n", pid)
				}
				if s3 == nil {
					s3 = aws.NewS3()
				}
				s3.Delete(pid)
			case CMD_AWS_TERMINATE_INSTANCE:
				if VERBOSE {
					log.Printf("Info: Terminating AWS EC2 instance '%s'\n", pid)
				}
				if ec2 == nil {
					ec2 = aws.NewEc2()
				}
				ec2.Add(pid) // Add instance ID to request
			default:
				// Remove using a command
				if DEBUG {
					log.Printf("Info: Cleanning up '%s' using command '%s'\n", pid, runCmds[cmd])
				}
				// Create command to be executed (or append)
				_, ok := runCmds[cmd]
				if ok {
					runCmds[cmd] = runCmds[cmd] + "\t" + pid
				} else {
					runCmds[cmd] = cmd + "\t" + pid
				}
		}
	}

	// Delete all EC2 instances
	if ec2 != nil {
		err := ec2.Terminate()
		if err != nil {
			log.Printf("Error: Terminating instances '%v', error: %s\n", ec2.InstanceIds, err)
		} else if VERBOSE {
			log.Printf("Info: Terminated AWS EC2 instances '%v'\n", ec2.InstanceIds)
		}
	}

	be.taskLoggerRunCmd(runCmds) // Run all commands (usually it's only one command)
}

/*
 Run all commands in the dictionary
*/
func (be *BdsExec) taskLoggerRunCmd(runCmds map[string]string) {
	for cmd, args := range runCmds {
		if len(cmd) > 0 {
			if VERBOSE {
				log.Printf("Info: Running command '%s'\n", cmd)
			}
			cmdExec := exec.Command(cmd)
			cmdExec.Args = strings.Split(args, "\t")
			err := cmdExec.Run()
			if err != nil {
				log.Fatal(err)
			}
		}
	}
}
